### PR TITLE
Extract separate components from FeaturesGrid

### PR DESCRIPTION
**Pull request by oswian**
Relates to https://github.com/Automattic/wp-calypso/issues/81078

## Proposed Changes

This PR follows up the work completed in https://github.com/Automattic/wp-calypso/pull/86730.  The focus here is on extracting components out of `../features-grid/index.tsx` into separate files within `../components/features-grid/**`. 

No change has been made to the internal implementation of these components.

## Testing Instructions

As this is a refactor there should be no functional changes. Build and tests should be 🟢.

Please smoke test as much as possible, focusing on the critical user flows.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by chriskmnds — CHANGES_REQUESTED** [review]
Thanks for working on this. I'll be looking more deeply and testing these thoroughly.

On the surface, I notice we are moving a lot of components into `/src/components/shared`, but it's not clear whether they are all indeed shared/needed outside of `FeaturesGrid`. 

`/src/components/shared` is meant to be a place for components that are shared between the main grid components (FeaturesGrid, ComparisonGrid, SpotlighPlan, PlanTypeSelector). Things like billing timeframe, actions, etc. If something is internal to `FeaturesGrid` then it should live under `/src/features-grid/`. 

At least that's the intention with the initial folder structure. Something like the following:
```
plans-grid-next/
├── components/
│   ├── comparison-grid/
│   │   ├── foo.tsx (component internal to comparison-grid)
│   │   └── index.tsx
│   ├── features-grid/
│   │   ├── bar.tsx (components internal to features-grid)
│   │   └── index.tsx
│   └── shared/
│       └── actions.tsx (component needed in both comparison-grid & features-grid)
├── ...
└── index.tsx
```

Can you confirm where we fall on that? Also, if something is indeed shared, then we should probably be updating its Prop type to not be extending `FeaturesGridProps` (where it makes sense). It might inform how better we can organize the overall types in the system i.e. if a set of props is needed for `PlanPrice` and that's a shared component between `FeaturesGrid` and `ComparisonGrid`, then it might be the case that `FeaturesGridProps` ought to extend from that instead of the other way around 🤔

**Comment by chriskmnds** [comment]
> if something is indeed shared, then we should probably be updating its Prop type to not be extending FeaturesGridProps (where it makes sense). It might inform how better we can organize the overall types in the system i.e. if a set of props is needed for PlanPrice and that's a shared component between FeaturesGrid and ComparisonGrid, then it might be the case that FeaturesGridProps ought to extend from that instead of the other way around 🤔

Of course, with a grain of salt i.e. only where it makes sense. Otherwise, not extending from anything might be better for shared components, rather than extending from "one of" the consumers' prop types.

**Comment by chriskmnds** [comment]
p.s. I updated the PR with:

1. Added to respective milestone https://github.com/Automattic/wp-calypso/milestone/455
2. Added to respective board (MarTech > Decepticons) as in-progress https://github.com/orgs/Automattic/projects/388
3. Added `Plans NPM: Essential` label
4. Assigned PR to you

these are all done through the sidebar

Screenshot 2024-01-25 at 1 18 53 PM

**Comment by chriskmnds** [comment]
> Thanks for working on this. I'll be looking more deeply and testing these thoroughly.
> 
> On the surface, I notice we are moving a lot of components into `/src/components/shared`, but it's not clear whether they are all indeed shared/needed outside of `FeaturesGrid`.
> 

Ohh I think these "shared" ones are between spotlight-card and features-grid, right?
It does complicate things a little when spotlight-card and features-grid are so similar, while more generic components also exist that are shared with the comparison-grid, like the existing `/src/components/billing-timeframe` or `/src/components/actions`. 

I wonder if instead we don't migrate spotlight-card to a separate folder and keep it inside `/src/features-grid`. Or maybe we do export and have it import components directly from under `/src/features-grid/**`. 🤔  Otherwise we do need some form of naming convention, which I sense will get complex. Not sure how else to have two "billing timeframe" components (one more generic and the other table-specific that we extract here)

**Comment by oswian** [comment]
Thanks for the review here @chriskmnds. Not quite ready for another review, however I've pushed an update which brings this up to speed with https://github.com/Automattic/wp-calypso/pull/86730 and addresses your feedback so far.

> Ohh I think these "shared" ones are between spotlight-card and features-grid, right?
It does complicate things a little when spotlight-card and features-grid are so similar, while more generic components also exist that are shared with the comparison-grid, like the existing /src/components/billing-timeframe or /src/components/actions.

Yes that's right. For this PR I've focused on extracting the shared components between `FeaturesGrid` and `SpotlightPlan` (based on my interpretation of https://github.com/Automattic/wp-calypso/issues/81078). Did you instead mean for this task to focus on extracting the shared components with `ComparisonGrid`?  I haven't (yet) reached the point of considering what components are shared with `ComparisonGrid`, however that feels like a good followup to this PR.

> I wonder if instead we don't migrate spotlight-card to a separate folder and keep it inside /src/features-grid. Or maybe we do export and have it import components directly from under /src/features-grid/**. 

In the latest update (https://github.com/Automattic/wp-calypso/pull/86825/commits/1f9482bf7993544d7502e249c865e29ea70a34ba) I've extracted `SpotlightPlan` into `/src/components/features-grid/spotlight-plan.tsx`. Both `FeaturesGrid` and `SpotlightPlan` import components from `/components/features-grid/**` such as:

```
import BillingTimeframes from './billing-timeframes';
import PlanHeaders from './plan-headers';
import PlanLogos from './plan-logos';
import PlanPrice from './plan-price';
import PlanStorageOptions from './plan-storage-options';
import PlanTagline from './plan-tagline';
import TopButtons from './top-buttons';
```


> /src/components/shared is meant to be a place for components that are shared between the main grid components (FeaturesGrid, ComparisonGrid, SpotlighPlan, PlanTypeSelector)

Yes that makes sense to me. In the latest update (https://github.com/Automattic/wp-calypso/pull/86825/commits/1f9482bf7993544d7502e249c865e29ea70a34ba), any components that have been extracted from `FeaturesGrid` now reside in `/components/features-grid/**` rather than `/components/shared`. In the followup mentioned above, if we find that some of these components are indeed shared with `ComparisonGrid` etc, they can be moved into `/components/shared`. Wdyt?

**Comment by oswian** [comment]
@chriskmnds Thinking further ahead, for when we add tests and storybook entries for these extracted components.

I see we have tests that live separately to the implementation, for example `BillingTimeframe`:

- Component: `src/components/shared/billing-timeframe/index.tsx`
- Test: `src/components/test/billing-timeframe.tsx`

What do you think about having tests live alongside the implementation:

Eg. 
- Component: `src/components/shared/billing-timeframe/index.tsx`
- Test: `src/components/shared/billing-timeframe/index.tests.tsx`
- Stories: `src/components/shared/billing-timeframe/index.stories.tsx`

Eg. 
- Hook: `src/hooks/use-is-large-currency.ts`
- Test: `src/hooks/use-is-large-currency.tests.ts`

I personally find this pattern easier to use as it avoids hunting around in other folders. It's easy to glance and see which components/hooks/other _don't_ have tests.

**Comment by oswian** [comment]
@chriskmnds Another question, this one related to how we approach styling.

Just for context, in recent times I've mostly relied on CSS modules to handle (most) scoping issues and style conflicts.

- `components/foo.tsx`
- `components/foo.scss`

```
import styles from "./foo.scss"

<div className={styles.hmm} />
```

I see in Calypso we instead use BEM (or something similar?) for scoping, and for `plans-grid-next` we have a single `src/style.scss` which appears to handle styling for every sub-component. 

I haven't been able to find any recent discussions around migrating to something like CSS modules, so I'll assume that's not currently an option. However is the plan to extract styles from `src/style.scss` so that each components defines it's own styling? So for example, alongside  `components/features-grid/plan-headers.tsx` we would create `components/features-grid/plan-headers.css` containing the relevant styling from `src/style.scss`:

```
.plan-features-2023-grid__header-title {
  ...
}
```

I see there's a precedent for this approach here:

https://github.com/Automattic/wp-calypso/blob/84eb10f36276e12b88f1188e1b539b0973ddf783/client/components/foldable-faq/index.tsx#L6

We don't get the benefit of CSS module's conflict avoidance, but it feels like a step in the right direction in terms of each component controlling their own styling.

**Comment by oswian** [comment]
@chriskmnds Opening this up for another review when you have time 🙏

**Comment by chriskmnds** [comment]
> @chriskmnds Thinking further ahead, for when we add tests and storybook entries for these extracted components.
> 
> I see we have tests that live separately to the implementation, for example `BillingTimeframe`:
> 
> * Component: `src/components/shared/billing-timeframe/index.tsx`
> * Test: `src/components/test/billing-timeframe.tsx`
> 
> What do you think about having tests live alongside the implementation:
> 
> ...
> 
> I personally find this pattern easier to use as it avoids hunting around in other folders. It's easy to glance and see which components/hooks/other _don't_ have tests.

Absolutely! Agreed. This pattern makes a lot of sense :-)

> We don't get the benefit of CSS module's conflict avoidance, but it feels like a step in the right direction in terms of each component controlling their own styling.

I think that makes a lot of sense too. Sometimes the `FeaturesGrid` or `ComparisonGrid` or Calypso (through `plans-features-main`) need to override styles, and that's something to happen at the respective ends (so some cleanup will be needed to migrate them out of the single `styles.scss` file).

There is a general trend towards `styled` components. However, a lot of code I've seen done in styled-components is not always up to standard IMO, a lot of times not done in any reusable manner or resolving to granular and random components that don't convey the right semantic/meaning. That's another thing to be a little more vigilant about.

For CSS class names, we should follow some pattern/convention. We do have some documentation (which I don't have handy right now), although with "Untangling Calypso" underway we probably ought to start sticking to .ORG guides more. What I propose is a form of BEM to prefix everything under `plans-grid-next` package with `.plans-grid-next__`. So features grid becomes `.plans-grid-next__features-grid`. A class for a header there becomes `.plans-grid-next__features-grid-header`, etc.

Let me know if you have other thoughts/suggestions.

File: packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		generatedWPComSubdomain,
+		hideUnavailableFeatures,
+		isCustomDomainAllowedOnFreePlan,
+		options,
+		paidDomainName,
+		renderedGridPlans,
+		selectedFeature,
+	} = props;
+	const translate = useTranslate();
+	const plansWithFeatures = renderedGridPlans.filter(
```
Comment (chriskmnds) [inline packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx]: Should we memoize this to maintain a single reference on rerenders?

File: packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		( gridPlan ) => ! isWpcomEnterpriseGridPlan( gridPlan.planSlug )
+	);
+
+	return (
+		<PlanFeaturesContainer
+			plansWithFeatures={ plansWithFeatures }
+			paidDomainName={ paidDomainName }
+			generatedWPComSubdomain={ generatedWPComSubdomain }
+			translate={ translate }
```
Comment (chriskmnds) [inline packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx]: We should get rid of the `translate` prop and grab from `useTranslate` in `PlanFeaturesContainer`.

File: packages/plans-grid-next/src/components/features-grid/mobile-free-domain.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { PlanFeaturesItem } from '../item';
+
+type MobileFreeDomainProps = {
+	gridPlan: GridPlan;
+	paidDomainName?: string;
+};
+
+const MobileFreeDomain = ( props: MobileFreeDomainProps ) => {
+	const { gridPlan, paidDomainName } = props;
+	const { planSlug, isMonthlyPlan } = gridPlan;
```
Comment (chriskmnds) [inline packages/plans-grid-next/src/components/features-grid/mobile-free-domain.tsx]: NIT. we could `const { gridPlan: { planSlug, isMonthlyPlan }, paidDomainName } = props`. or better yet, destructure in the signature

File: packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+const PlanFeaturesList = ( props: PlanFeaturesListProps ) => {
+	const {
+		generatedWPComSubdomain,
+		hideUnavailableFeatures,
+		isCustomDomainAllowedOnFreePlan,
+		options,
+		paidDomainName,
+		renderedGridPlans,
+		selectedFeature,
+	} = props;
```
Comment (chriskmnds) [inline packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx]: We can expand directly in the signature:

```
const PlanFeaturesList = ( { generatedWPComSubdomain, ... }: PlanFeaturesListProps ) => { ...
```

Do you prefer to map onto `props` and destructure from there?

File: packages/plans-grid-next/src/components/actions.tsx [hunk 5]
```diff
@@ -528,14 +528,14 @@ const PlanFeaturesActionsButton: React.FC< PlanFeaturesActionsButtonProps > = (
 	);
 };
 
-const PlanFeatures2023GridActions = ( props: PlanFeaturesActionsButtonProps ) => {
+const PlanFeatures2023GridActions = localize( ( props: PlanFeaturesActionsButtonProps ) => {
```
Comment (chriskmnds) [inline packages/plans-grid-next/src/components/actions.tsx]: Is this necessary (`localize`)? 🤔

**Review by chriskmnds — APPROVED** [review]
Thanks! works for me. I haven't noticed any regressions. I left some comments in the code, but nothing critical :-)

File: packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx (hunk 2 above)
Comment (oswian) [reply packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx]: Good catch, updated:https://github.com/Automattic/wp-calypso/pull/86825/commits/306deb0ea705e6216562802758675a5d966accd4

File: packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx (hunk 4 above)
Comment (oswian) [reply packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx]: > Do you prefer to map onto props and destructure from there?

I find this approach can look a bit cleaner when there are a lot of props. Happy to destructure in the signature too though, particularly if that's more common in Calypso.

Updated: [7aff101](https://github.com/Automattic/wp-calypso/pull/86825/commits/7aff101b05371994aa1835db2d911777a224a9ff)

File: packages/plans-grid-next/src/components/features-grid/mobile-free-domain.tsx (hunk 3 above)
Comment (oswian) [reply packages/plans-grid-next/src/components/features-grid/mobile-free-domain.tsx]: 👍  Updated: [8bf9239](https://github.com/Automattic/wp-calypso/pull/86825/commits/8bf9239eea34735b0fcc5752cce91a598abb61d9)

File: packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx (hunk 1 above)
Comment (oswian) [reply packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx]: Good idea, updated: https://github.com/Automattic/wp-calypso/pull/86825/commits/ec922c37d021fab247181d156be1898c86451518

File: packages/plans-grid-next/src/components/actions.tsx (hunk 5 above)
Comment (oswian) [reply packages/plans-grid-next/src/components/actions.tsx]: Good question.

This change resolved a TS issue I encountered in vscode, where the prop types were not resolving correctly in components that imported `PlanFeatures2023GridActions`. However now when I revert this change, I no longer encounter the same issue 🤔 

So as the first step, I'll revert this change. 

In terms of removing the call to `localize(...)` completely from this component, yes I believe we can. From the [docs](https://github.com/Automattic/wp-calypso/tree/trunk/packages/i18n-calypso#react-hook) I see that  `useTranslate` was introduced as the modern alternative to `localize`. We're already using `useTranslate` in this component, so the use of `localize` is redundant.

Updated: [36a521c](https://github.com/Automattic/wp-calypso/pull/86825/commits/36a521c695947d407b44974385318f6fda4e107a)

**Comment by oswian** [comment]
Thanks for the feedback @chriskmnds. I've addressed all of your comments.

**Comment by oswian** [comment]
@chriskmnds Just following up on these comments:

> There is a general trend towards styled components. However, a lot of code I've seen done in styled-components is not always up to standard IMO, a lot of times not done in any reusable manner or resolving to granular and random components that don't convey the right semantic/meaning. That's another thing to be a little more vigilant about.

Thanks for the heads up regarding the use of styled components. I have only dabbled with styled components, and it was quite a while ago. Regarding the general trend _towards_ styled components - is that trend at the Calypso project level, or our team? Is the current plan to continue using a mix of .css files alongside styled components, and defer to each dev's preference when implementing new components? 

> For CSS class names, we should follow some pattern/convention. We do have some documentation (which I don't have handy right now), although with "Untangling Calypso" underway we probably ought to start sticking to .ORG guides more. What I propose is a form of BEM to prefix everything under plans-grid-next package with .plans-grid-next__. So features grid becomes .plans-grid-next__features-grid. A class for a header there becomes .plans-grid-next__features-grid-header, etc.

This makes sense to me 👍

**Comment by chriskmnds** [comment]
> Thanks for the heads up regarding the use of styled components. I have only dabbled with styled components, and it was quite a while ago. Regarding the general trend towards styled components - is that trend at the Calypso project level, or our team? Is the current plan to continue using a mix of .css files alongside styled components, and defer to each dev's preference when implementing new components?

I am not sure if it's a general trend, or something that's stabilized or becoming convention/recommendation of sorts. (CSS is fine by me, and I personally much prefer it over complex/ambiguous styled-components just for the sake of it.) - but make a best judgment - I think a good design/abstraction will work either way!

There is also been, interestingly, some exploration by team Calypso to move away from `Emotion`. Just read this in a team update (p9oQ9f-27R-p2):

> **Explore alternatives to Emotion**
>
> Emotion has received some criticism in the past, mainly regarding the increase to bundle size and the added computational overhead in the browser.
> 
> We could explore a few alternatives that would still ensure style encapsulation:
> 
> - Sass or CSS modules
> - [StyleX](https://href.li/?https://stylexjs.com/docs/learn/)
> - [Vanilla Extract](https://href.li/?https://vanilla-extract.style/)
> - [Panda CSS](https://href.li/?https://panda-css.com/)
> - [Stitches](https://href.li/?https://stitches.dev/)
> - …

Not surprising tbh. No idea at what level and with what momentum that's happening though.

**Comment by oswian** [comment]
Interesting, thanks for link @chriskmnds.

**Pull request merged**